kernel/posix/mqueue.c

```diff
@@ -0,0 +1,532 @@
+		    attrs->mq_maxmsg > CONFIG_MSG_COUNT_MAX) {
+			error = true;
+			goto done;
+		}
+
+		/* Allocate space for message queue object */
+		if (k_mem_pool_alloc(&posix_mqueue_pool, &block_memq_obj,
+				     sizeof(struct mqueue_object), 100) == 0) {
+			memset(block_memq_obj.data, 0,
+			       sizeof(struct mqueue_object));
+			msg_queue = (struct mqueue_object *)block_memq_obj.data;
+			memcpy(&msg_queue->block_mqueue, &block_memq_obj,
+			       sizeof(struct k_mem_block));
+		} else {
+			error = true;
+			goto done;
+		}
+
+		/* Allocate space for block tracking */
+		if (k_mem_pool_alloc(&posix_mqueue_pool, &block_track,
+				     sizeof(struct mem_block),
+				     K_FOREVER) == 0) {
+			memset(block_track.data, 0, sizeof(struct mem_block));
+			msg_queue->block_buffer_name = block_track.data;
+		} else {
+			error = true;
+			goto done;
+		}
+
+		/* Allocate space for message queue name */
+		if (k_mem_pool_alloc(&posix_mqueue_pool, &block_memq_name,
+				     strlen(name) + 1, K_FOREVER) == 0) {
+			memset(block_memq_name.data, 0, strlen(name)+1);
+			msg_queue->name = block_memq_name.data;
+			memcpy(&msg_queue->block_buffer_name->block_name,
+			       &block_memq_name, sizeof(struct k_mem_block));
+		} else {
+			error = true;
+			goto done;
+		}
+
+		strcpy(msg_queue->name, name);
+
+		msg_size = attrs->mq_msgsize;
+		max_msgs = attrs->mq_maxmsg;
+
+		/* Allocate space for message queue buffer */
+		if (k_mem_pool_alloc(&posix_mqueue_pool, &block_buffer,
+				     msg_size * max_msgs * sizeof(u8_t), 100)
+				     == 0) {
+			memset(block_buffer.data, 0,
+			       msg_size * max_msgs * sizeof(u8_t));
+			buffer = block_buffer.data;
+			memcpy(&msg_queue->block_buffer_name->block_buffer,
+			       &block_buffer, sizeof(struct k_mem_block));
+		} else {
+			error = true;
+			goto done;
+		}
+
+		msg_queue->ref_count = 1;
+		/* initialize zephyr message queue */
+		k_msgq_init(&msg_queue->queue, buffer, msg_size, max_msgs);
+
+		/* Add the message queue at start of list */
+		k_sem_take(&mq_sem, K_FOREVER);
+		msg_queue->prev = NULL;
+
+		if (mq_list != NULL) {
+			mq_list->prev = msg_queue;
+		}
+
+		msg_queue->next = mq_list;
+		mq_list = msg_queue;
+	} else {
+		k_sem_take(&mq_sem, K_FOREVER);
+
+		if (msg_queue->ref_count >= CONFIG_MAX_DESC_PERMQ) {
+			k_sem_give(&mq_sem);
+			errno = ENOSPC;
+			k_mem_pool_free(&block_memq_desc);
+			return (mqd_t)mqd;
+		}
+
+		msg_queue->ref_count += 1;
+	}
+
+	k_sem_give(&mq_sem);
+	msg_queue_desc->mqueue = msg_queue;
+	msg_queue_desc->flags = (oflags & O_NONBLOCK) ? O_NONBLOCK : 0;
+
+done:
+	if (error) {
+		errno = ENOSPC;
+		if (msg_queue != NULL) {
+			if (msg_queue->block_buffer_name != NULL) {
+				k_mem_pool_free(&block_track);
+			}
+
+			if (msg_queue->name != NULL) {
+				k_mem_pool_free(&block_memq_name);
+			}
+
+			k_mem_pool_free(&block_memq_obj);
+		}
+
+		if (msg_queue_desc != NULL) {
+			k_mem_pool_free(&block_memq_desc);
+		}
+	} else {
+		mqd = msg_queue_desc;
```

Comment:
Why does this assignment exist?  Why not return msg_queue_desc directly?

Reply:
Yes we can avoid it. I was keeping for uniformity at return. We already have an variable mqd.